docs/internal-documentation/templates/specifications/tests/Template01.md

```diff
@@ -0,0 +1,253 @@
+> > such as the one in the summary section.
+
+# Template01: This is a test specification template
+
+> > Replace "Template01" with test case ID. Replace the text with at short
+> > description
+
+## Test case identifier
+**Template01**
+
+> > Replace with correct test case ID as specified in the
+> > [Test Case Identifier Specification].
+
+## Table of contents
+
+> > If the specification contains extra sections, or if some section is not
+> > included, update the list. In the normal case, keep the following sections.
+
+* [Objective](#Objective)
+* [Scope](#Scope)
+* [Inputs](#Inputs)
+* [Summary](#Summary)
+* [Test procedure](#Test-procedure)
+* [Outcome(s)](#Outcomes)
+* [Special procedural requirements](#Special-procedural-requirements)
+* [Intercase dependencies](#Intercase-dependencies)
+* [Terminology](#terminology)
+
+## Objective
+
+> > Write a description of what this test case tests, i.e. the details that are
+> > considered to be right and wrong. Include references to RFCs and other
+> > documents.
+> >
+> > Do not include full links here (in the source). Put them at the end of the
+> > specification document instead, as in this template.
+> >
+> > Do have deep links, but keep the display text short if possible. E.g.:
+
+...described in [RFC 4035][RFC 4035#section-3.1.3], section 3.1.3, ...
+
+## Scope
+
+> > If this test case assumes, but not depends on, another test case then specify
+> > it. Dependency would mean that the test case is affected by the other test
+> > case, or even cannot be run if the other has not been run. Assuming that
+> > another test case has been run is just to make it possible to ignore certain
+> > errors. E.g.:
+
+It is assumed that *Child Zone* has been tested and reported by [Basic04]. This
+test case will just ignore non-responsive name servers or name servers not
+giving a correct DNS response for an authoritative name server.
+
+
+## Inputs
+
+> > Input data to the test case. Always include the Child zone, but it can also
+> > be other data units.
+> >
+> > The input data name is put in quote marks '""', then a hyphen '-' as a
+> > separator, and finally a description.
+
+* "Child Zone" - The domain name to be tested.
+
+## Summary
+
+> > First we can have bullets, if applicable, that states noteable things about
+> > the execution or the messages. E.g.
+* If no CDS record is found, the test case will terminate early
+  with no message tag outputted.
+* If a CDS record is of "delete" type, then it can by definition not
+  match or point at any DNSKEY record.
+
+> > Here is a table of all message tags referred to in the steps. The tag in the
+> > first column, the default severity level in the second, and a statement on
+> > when the message is outputted in the third. Always use the same table set-up,
+> > but with the correct tags. E.g.:
+
+Message Tag outputted         | [Default level] | Description of when message tag is outputted
```

Comment:
Do you think it can be a good idea to specify upfront if a message tag uses any parameters (as a opposed to only putting them in the test procedure for instance)? I was reading #942 and I think that can help know what to expect in the test procedure.

Reply:
That is a very good idea. I will update the template to meet that.

Reply:
I think that the idea is good, but complex. I thought of having more information, but decided to start with just indicating that that message will be accompanied by data.

 The messages can have many parameters. The following is many an extreme example with five parameters (it was one of the first in the PO file). In most cases there are probably not more than three, but the model must also work for the extreme cases.
```
#. ADDRESS:NAMESERVER_IP_PRIVATE_NETWORK
"Nameserver {nsname} has an IP address ({ns_ip}) with prefix {prefix} "
"referenced in {reference} as a '{name}'."
```
The parameter names are data types and defined in [logentry_args] (needs update). Is it the data types that should go into the specification?

One consequence is that the raw markdown tables will be hard to read.


[logentry_args]: https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md

Reply:
Maybe we can put just a coma separated list of parameter names, at least it gives some rough information about what is expected. If the data types are explained more in details elsewhere, even if it's currently outdated, that can be sufficient.

Reply:
We can try that model. It will require that [logentry_args] is updated to be complete. It will probably require that the parameter (argument) name are more explaining. "Integer" could work in a msgid but it does not work will in a test case specification.

[logentry_args]: https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md

Reply:
>  It will probably require that the parameter (argument) name are more explaining. "Integer" could work in a msgid but it does not work will in a test case specification.

I agree that parameter names need to be explicit, and actually even in msgid I think it is nice to have explicit parameters to ease the work of translators.

Reply:
It is a balance to keep them short and to make them understandable. In most cases there is enough context in the msgid so that you can see what it is if there is information on what the data is, which "logentry_args" gives if the argument name is documented there.